Add usage_type, aggregate_usage, and billing_scheme.

blvrd opened the pull request:
<!--
  Please give us ~1 week to get back to you.
  If you'd like to receive occasional updates, sign up for our newsletter at http://tinyletter.com/stripe-rails
-->

After trying to create a metered plan using the `usage_type` configuration parameter, I noticed that a `licensed` plan was actually created in Stripe. It turns out that a few options were not being passed through, including `usage_type`, `aggregate_usage`, and `billing_scheme`, when pulling together the data to send to Stripe.

This now works for me, but I'm not sure the best way to test the new implementation. @tansengming how would you approach testing it?

Commit:
Add usage_type, aggregate_usage, and billing_scheme.

tansengming committed:
Merge branch 'master' into usage-type

tansengming commented:
Thanks for the Pull Request @garrettqmartin8 . To test this I would add a new test just for creating metered plans over at https://github.com/tansengming/stripe-rails/blob/master/test/plan_builder_spec.rb


try to make sure that the correct params are sent to the Stripe API like this:

https://github.com/tansengming/stripe-rails/blob/6bf308b51dd65e6fd08084a1e21fdcd40701e1f5/test/plan_builder_spec.rb#L213-L224

Commit:
Add tests for metered plans

blvrd commented:
@tansengming thanks for pointing me in the right direction. I've added a test case and a new plan fixture. Let me know if there's anything I'm missing to make this mergeable.

tansengming commented:
Nice work! Looks like there are issues with code climate can you have a look at it please? Thank you!

Commit:
DRY up create options

blvrd commented:
I fixed a couple, but the two that remain are due to some minor duplication across two different files. Not sure of a good way to refactor that.

tansengming commented:
no worries, I'll try to fix it.

tansengming commented on lib/stripe/plans.rb:
```diff
@@ -80,20 +79,14 @@ def product_options
-          billing_scheme: billing_scheme,
-          statement_descriptor: statement_descriptor
-        }.delete_if{|_, v| v.nil? }
+        default_create_options.merge(product_options)
```
though unlikely, I'm a bit worried that `product_id` (see code snippet below) might leak into this options hash so I'm going to revert this commit.

https://github.com/tansengming/stripe-rails/blob/6bf308b51dd65e6fd08084a1e21fdcd40701e1f5/lib/stripe/plans.rb#L76-L78

blvrd replied:
Ah, I see. No problem!

tansengming committed:
Revert "DRY up create options"

This reverts commit b20a2dd6ae1861b10e4d2df77ecd589d3e6fb814.

tansengming committed:
DRY up `.delete_if { |_, v| v.nil? }` with a new method

tansengming commented on lib/stripe/plans.rb:
```diff
@@ -86,8 +89,11 @@ def create_options_without_products
           interval_count: interval_count,
           trial_period_days: trial_period_days,
           metadata: metadata,
+          usage_type: usage_type,
```
I'm removing these lines since these are the deprecated options for the old API and do not need to be updated. I'll make a note on the method.

Unfortunately the [Stripe API reference](https://stripe.com/docs/api) only shows me the latest version of the API so I cannot confirm that these are actually invalid.

blvrd replied:
That makes sense.

tansengming committed:
reverts changes to create_options_without_products,

since they are for an older API and probably do not support the new options.
Also added a note up top.

tansengming committed:
fixes spec

tansengming commented:
🍾 

@garrettqmartin8 This looks good to go! I'll try to get this pushed to ruby gems soon so you'll be able to enjoy the fruits of your labor. Thanks so much for fixing this.

blvrd commented:
Great, thanks @tansengming. Looking forward to contributing more.
